src/main/java/pl/damianszczepanik/jenkins/buildhistorymanager/model/actions/DeleteArtifactsMatchingPatternsAction.java

```diff
@@ -0,0 +1,168 @@
+package pl.damianszczepanik.jenkins.buildhistorymanager.model.actions;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+
+import org.jenkinsci.remoting.RoleChecker;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+import hudson.FilePath;
+import hudson.FilePath.FileCallable;
+import hudson.Util;
+import hudson.model.Run;
+import hudson.remoting.VirtualChannel;
+import jenkins.util.VirtualFile;
+
+/**
+ * Deletes the artifacts with patterns.
+ */
+public class DeleteArtifactsMatchingPatternsAction extends Action {
+    private static final Logger LOG = Logger.getLogger(DeleteArtifactsMatchingPatternsAction.class.getName());
+
+    private String includePatterns;
+    private String excludePatterns;
+
+    @DataBoundConstructor
+    public DeleteArtifactsMatchingPatternsAction() {
+        // Jenkins stapler requires to have public constructor with @DataBoundConstructor
+    }
+
+    public String getIncludePatterns() {
+        return includePatterns;
+    }
+
+    @DataBoundSetter
+    public void setIncludePatterns(String includePatterns) {
+        this.includePatterns = includePatterns;
+    }
+
+    public String getExcludePatterns() {
+        return excludePatterns;
+    }
+
+    @DataBoundSetter
+    public void setExcludePatterns(String excludePatterns) {
+        this.excludePatterns = excludePatterns;
+    }
+
+    static boolean isDirectoryEmpty(Path path) throws IOException {
+        try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(path)) {
+            return !directoryStream.iterator().hasNext();
+        }
+    }
+
+    // if 'file' is on a different node, this FileCallable will be transferred to that node and executed there.
+    public static final class Delete implements FileCallable<Void> {
+        private static final long serialVersionUID = -7712623306942117400L;
+        private final File archiveRootFile;
+
+        public Delete(File archiveRootFile) {
+            this.archiveRootFile = archiveRootFile;
+        }
+
+        @Override 
+        public Void invoke(File file, VirtualChannel channel) throws IOException {
+            Set<File> directories = new HashSet<>();
+            deleteFileOrLogError(file, directories);
+            deleteEmptyDirectoriesAndParents(directories);
+            return null;
+        }
+
+        void deleteEmptyDirectoriesAndParents(Set<File> directories) throws IOException {
+            for (File dir : directories) {
+                if (shouldDeleteDirectory(dir)) {
```

Comment:
Based what I see in the code this loop will be iterated only once so having collection at this level makes no sense

Reply:
what about this ?

Reply:
You are right. Thanks for the feedback. I've addressed it by removing the unnecessary collection of directories from the code and updated the associated methods and tests.
 
Here is the commit:
https://github.com/jenkinsci/build-history-manager-plugin/pull/84/commits/017b0af439a63782711ab0d10200b6b1652d3149